File: 2831-Find-the-Longest-Equal-Subarray/main.go
package main

import "fmt"

func main() {
	testArr1 := []int{1, 3, 2, 3, 1, 3}
	kArr1 := 3

	testArr2 := []int{1, 1, 2, 2, 1, 1}
	kArr2 := 2

	fmt.Printf("Solution 1: %d\n", longestEqualSubarray(testArr1, kArr1))
	fmt.Printf("Solution 2: %d\n", longestEqualSubarray(testArr2, kArr2))
}

func longestEqualSubarray(nums []int, k int) int {

	n := len(nums)
	j, res, cnt := 0, 0, 0

	freq := make(map[int]int)
	for i := 0; i < n; i++ {
		freq[nums[i]] += 1

		cnt = max(res, freq[nums[i]])

		if cnt+k < i-j+1 {
			freq[nums[j]]--
			j++
		} else {
			res = max(res, cnt)
		}
	}

	return res
}

func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}

File: 2831-Find-the-Longest-Equal-Subarray/main.py
from typing import List
from collections import defaultdict
"""

You are given a 0-indexed integer array nums and an integer k.

A subarray is called equal if all of its elements are equal. Note that the empty subarray is an equal subarray.

Return the length of the longest possible equal subarray after deleting at most k elements from nums.

A subarray is a contiguous, possibly empty sequence of elements within an array.

For each number x in nums, create a sorted list indicesx of all indices i such that nums[i] == x.
On every indicesx, execute a sliding window technique.
For each indicesx, find i, j such that (indicesx[j] - indicesx[i]) - (j - i) <= k and j - i + 1 is maximized.
The answer would be the maximum of j - i + 1 for all indicesx.

Example 1:

Input: nums = [1,3,2,3,1,3], k = 3
Output: 3
Explanation: It's optimal to delete the elements at index 2 and index 4.
After deleting them, nums becomes equal to [1, 3, 3, 3].
The longest equal subarray starts at i = 1 and ends at j = 3 with length equal to 3.
It can be proven that no longer equal subarrays can be creat
"""

class Solution:
    def longestEqualSubarray(self, nums: List[int], k: int) -> int:

        n = len(nums)
        left, res = 0, 0

        freq = defaultdict(int)
        cnt = 0

        for right in range(n):
            freq[nums[right]] += 1

            cnt = max(cnt, freq[nums[right]])

            if cnt + k < right - left + 1:
                freq[nums[left]] -= 1
                left +=1
            else:
                res = max(res,  cnt)

        return res
    

s = Solution()

print(s.longestEqualSubarray([1,3,2,3,1,3], 3), "solution 1")
print(s.longestEqualSubarray([1,1,2,2,1,1], 2), "solution 2")
